### ISSUE TITLE
Side load embedded entities in REST requests into Core Data store

### ISSUE THREAD

**Issue by getdave** [WordPress/gutenberg#49916]
The REST API supports [Linking and Embedding](https://developer.wordpress.org/rest-api/using-the-rest-api/linking-and-embedding/). This is allows embedding related resources together in one response.

For example, if you requested all Posts `wp-json/wp/v2/posts` by default each Post in the response includes a `_links` field which provides information about where you can request the full User object for the Post's Author.

```
"_links": {
    "author": [
        {
            "embeddable": true,
            "href": "http://localhost:8888/wp-json/wp/v2/users/1"
        }
    ],
}
```

Note however, that the full User object is not included in the response. If you wish to include that then you can pass the `_embed` parameter `wp-json/wp/v2/posts?_embed`. The response will then include the full User object for the `author` under the `_embedded` key.

We know when a given field is embeddable as that is indicated by the `"embeddable": true`.

## Suggestion

Currently `@worpress/core-data` doesn't seem to support sideloading entities embedded in a given response. We should consider adding this as an optional feature in order to reduce the need for additional queries.

To do this we need to support:

- ability to add `_embed` query param
- add any entities found under the `_embedded` key into the state

**Comment by getdave** [comment]
cc @spacedmonkey @TimothyBJacobs who suggested this to me.

**Comment by youknowriad** [comment]
Thanks for opening the issue, embedding is definitely something we could leverage more in core-data. I do feel that it should remain an internal implementation detail though and not leak into the public API of `core-data` though. So I think it's better to approach this with a use-case in mind and see how to address that use-case instead.

If I'm not wrong, you're trying to load the "fallback" navigation menu using this technique. So the public API is actually something like  `wp.data.select('core').getFallbackNavigationMenu()` right? I'm curious to know how you're addressing that use-case at the moment (without embedding support), what's the public API you're using (I'm guessing two selector calls)?

**Comment by getdave** [comment]
> So I think it's better to approach this with a use-case in mind and see how to address that use-case instead.

Yep I was trying to find one other than Nav fallbacks, but couldn't think of one yet.

> ...an internal implementation detail though and not leak into the public API of core-data though

You're suggesting avoiding exposing the need to pass `embed` as an arg for example. Rather Core Data should just handle it for you. My only thought was that embedding is more "expensive" and thus its likely we'd want to be selective in when resources are embedded. If we can achieve that programmatically then great, but I wasn't sure how.

This came out of the fallbacks work but we're not going to attempt to address it for that PR to land as it's not essential to the implementation.

> If I'm not wrong, you're trying to load the "fallback" navigation menu using this technique

Currently we're _not_ using Core Data for the fallbacks, rather direct fetch requests. I wasn't entirely sure it _should_ be part of the Core Data API (seemed Nav block specific) but we can easily look at that after the fact.

> I'm guessing two selector calls

Yes that's basically correct (technical implementation details aside).

If we were able to:

- include a Core Data selector for getting fallbacks
- support embedding

...then this would reduce a _single_ additional network request on the Nav block the _first time_ it loads in the editor.

**Comment by youknowriad** [comment]
> include a Core Data selector for getting fallbacks

I think if you include a selector to core data, you'll need a reducer to track the "fallback nav id" in the state.
you could use the "embed" request in the resolver of that selector to only make a single request and when you receive the response, you can update the nav id in the reducer and at the same time, you fill the "entities" (receiveEntities). 

I think it's very hard to talk in abstract here, rather, I'd prefer for you as a consumer to think of what kind of API you'd want to use in your block ideally.

**Comment by getdave** [comment]
As a consumer in this specific instance I'd want to call

```
wp.data.select('core').getFallbackNavigationMenu()
```

...and it would

- return the menu ID
- sideload the entity into state

That way when the block makes a `wp.data.select('core').getEntityRecord('postType','wp_navigation', SOME_ID)` the data would be in state already and no network request would be dispatched.

How `getFallbackNavigationMenu` does that under the hood is a lower level detail. 

As a developer-consumer of Core Data, implementing the `getFallbackNavigationMenu` selector I'd expect to be able to manually decide whether or not the data is sideloaded (as this depends on the use case for my selector).

**Comment by youknowriad** [comment]
`getFallbackNavigationMenu` sounds like it should return the object `getFallbackNavigationMenuId` is more something that should return an ID. We can implement them internally in anyway we want and share the same resolver if needed.

**Comment by getdave** [comment]
I experimented with this approach a bit today. In short it works. I'll get a PoC PR up once https://github.com/WordPress/gutenberg/pull/48698 is merged.

**Comment by kadamwhite** [comment]
> My only thought was that embedding is more "expensive" and thus its likely we'd want to be selective in when resources are embedded.

Strong +1 to Dave's concern here, I would discourage over-use of embedding. It's "easy", but not efficient. We can be much more targeted about what data you get by making multiple requests, and there are no controls equivalent to `_fields` which can be used to opt-out of processing of embedded resources, so parallel requests themselves will be less draining on server resources than one big _embed. If the goal is to have core data orchestrate that for you, I would think that could be accomplished with parallel or serial requests instead of having to embed.

**Comment by getdave** [comment]
Thanks @kadamwhite. What I'm advocating for is that we allow the _developer_ to _opt in_ to embedding at their own discretion. 

That allows them to weigh the tradeoffs. 

In the context of the Navigation block I believe the additional request time on a _single_ (embedded) request that allows us to pre-populate the data store with the retrieved Navigation entity and thus avoid a _2nd_ network request would definitely be worthwhile. However in other scenarios this may not be the case.

**Comment by getdave** [comment]
[The fallbacks work has now merged](https://github.com/WordPress/gutenberg/pull/48698). The discussion about Nav block specific sideloading can continue in https://github.com/WordPress/gutenberg/issues/50002.

**Comment by Mamaduka** [comment]
> ability to add _embed query param

Just like `_fields`, this has been supported for a while now, and core already uses it in some places. The following query will return the embedded author:

```js
wp.data.select( 'core' ).getEntityRecord( 'postType', 'post', 23, { _embed: 'author' } )
```

> add any entities found under the _embedded key into the state

Similar to some existing optimizations, I'm guessing we want to pre-resolve calls to the embedded entity record, if it's already requested elsewhere in the tree.

I'm not sure if that's doable. IIRC, the embedded data will be returned for a different `context` than the original requests. Just like `edit` and `view` context, there's also an `embed` context just for embeddables.

Maybe it's better to leave this up to consumers like @youknowriad suggested.

**Example**

```js
// This is triggered first.
wp.data.select( 'core' ).getEntityRecord( 'postType', 'post', 23, { _embed: 'author' } )

// Assuming previously embedded post author's ID is 1, this won't trigger a resolution.
wp.data.select( 'core' ).getEntityRecord( 'root', 'user', 1 )
```
